### ISSUE TITLE
example panic

### ISSUE THREAD

**Issue by icexin** [inkyblackness/imgui-go#94]
example `example_glfw_opengl3`

![console](https://user-images.githubusercontent.com/2713364/79444619-ff0a0b80-800d-11ea-887b-b05e73308948.gif)

The program hangs when the window is closed.

panic stack
```
fatal error: exitsyscall: syscall frame is no longer valid

goroutine 1 [running, locked to thread]:
runtime.throw(0x41f2f2a, 0x2d)
	/Users/fanbingxin/local/go/src/runtime/panic.go:774 +0x72 fp=0xc000048e48 sp=0xc000048e18 pc=0x402c562
runtime.exitsyscall()
	/Users/fanbingxin/local/go/src/runtime/proc.go:2957 +0x27d fp=0xc000048e78 sp=0xc000048e48 pc=0x403519d
runtime.cgocallbackg(0x0)
	/Users/fanbingxin/local/go/src/runtime/cgocall.go:188 +0xa3 fp=0xc000048ee0 sp=0xc000048e78 pc=0x4004e83
runtime.cgocallback_gofunc(0x40474ec, 0x4057dc0, 0xc000048f60, 0xc000048f90)
	/Users/fanbingxin/local/go/src/runtime/asm_amd64.s:793 +0x9b fp=0xc000048f00 sp=0xc000048ee0 pc=0x4055e5b
runtime.asmcgocall(0x4057dc0, 0xc000048f60)
	/Users/fanbingxin/local/go/src/runtime/asm_amd64.s:640 +0x42 fp=0xc000048f08 sp=0xc000048f00 pc=0x4055cf2
runtime.libcCall(0x4057dc0, 0xc000048f60, 0xc0000240b8)
	/Users/fanbingxin/local/go/src/runtime/sys_darwin.go:46 +0x6c fp=0xc000048f38 sp=0xc000048f08 pc=0x40474ec
runtime.exit(0xc000000000)
	/Users/fanbingxin/local/go/src/runtime/sys_darwin.go:220 +0x31 fp=0xc000048f60 sp=0xc000048f38 pc=0x4047aa1
runtime.main()
	/Users/fanbingxin/local/go/src/runtime/proc.go:225 +0x272 fp=0xc000048fe0 sp=0xc000048f60 pc=0x402df42
runtime.goexit()
	/Users/fanbingxin/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc000048fe8 sp=0xc000048fe0 pc=0x4056581
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
panic during panic
SIGABRT: abort
PC=0x7fff60d7eb66 m=0 sigcode=0

goroutine 0 [idle]:
runtime.exit(0xc000000000)
	/Users/fanbingxin/local/go/src/runtime/sys_darwin.go:220 +0x31 fp=0xc000048f60 sp=0xc000048f38 pc=0x4047aa1
runtime.main()
	/Users/fanbingxin/local/go/src/runtime/proc.go:225 +0x272 fp=0xc000048fe0 sp=0xc000048f60 pc=0x402df42
runtime.goexit()
	/Users/fanbingxin/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc000048fe8 sp=0xc000048fe0 pc=0x4056581

goroutine 1 [running, locked to thread]:
runtime.systemstack_switch()
	/Users/fanbingxin/local/go/src/runtime/asm_amd64.s:330 fp=0xc000048de0 sp=0xc000048dd8 pc=0x40544d0
runtime.fatalthrow()
	/Users/fanbingxin/local/go/src/runtime/panic.go:826 +0x57 fp=0xc000048e18 sp=0xc000048de0 pc=0x402c737
runtime.throw(0x41f2f2a, 0x2d)
	/Users/fanbingxin/local/go/src/runtime/panic.go:774 +0x72 fp=0xc000048e48 sp=0xc000048e18 pc=0x402c562
runtime.exitsyscall()
	/Users/fanbingxin/local/go/src/runtime/proc.go:2957 +0x27d fp=0xc000048e78 sp=0xc000048e48 pc=0x403519d
runtime.cgocallbackg(0x0)
	/Users/fanbingxin/local/go/src/runtime/cgocall.go:188 +0xa3 fp=0xc000048ee0 sp=0xc000048e78 pc=0x4004e83
runtime.cgocallback_gofunc(0x40474ec, 0x4057dc0, 0xc000048f60, 0xc000048f90)
	/Users/fanbingxin/local/go/src/runtime/asm_amd64.s:793 +0x9b fp=0xc000048f00 sp=0xc000048ee0 pc=0x4055e5b
runtime.asmcgocall(0x4057dc0, 0xc000048f60)
	/Users/fanbingxin/local/go/src/runtime/asm_amd64.s:640 +0x42 fp=0xc000048f08 sp=0xc000048f00 pc=0x4055cf2
runtime.libcCall(0x4057dc0, 0xc000048f60, 0xc0000240b8)
	/Users/fanbingxin/local/go/src/runtime/sys_darwin.go:46 +0x6c fp=0xc000048f38 sp=0xc000048f08 pc=0x40474ec
runtime.exit(0xc000000000)
	/Users/fanbingxin/local/go/src/runtime/sys_darwin.go:220 +0x31 fp=0xc000048f60 sp=0xc000048f38 pc=0x4047aa1
runtime.main()
	/Users/fanbingxin/local/go/src/runtime/proc.go:225 +0x272 fp=0xc000048fe0 sp=0xc000048f60 pc=0x402df42
runtime.goexit()
	/Users/fanbingxin/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc000048fe8 sp=0xc000048fe0 pc=0x4056581

rax    0x0
rbx    0x7fff99676380
rcx    0x7ffeefbff328
rdx    0x0
rdi    0x307
rsi    0x6
rbp    0x7ffeefbff360
rsp    0x7ffeefbff328
r8     0x0
r9     0x0
r10    0x0
r11    0x206
r12    0x307
r13    0x4
r14    0x6
r15    0x2d
rip    0x7fff60d7eb66
rflags 0x206
cs     0x7
fs     0x0
gs     0x0
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
stack trace unavailable
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
example_glfw_opengl3(4340,0x7fff99676380) malloc: *** error for object 0x451a0c0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
fatal: morestack on gsignal
```

**Comment by dertseha** [comment]
Hello there and thank you for your message!
I currently don't have access to a Mac development environment, so I'd need help figuring out what is wrong here.

My guiding questions: Does this also happen with the SDL example? If it is GLFW-only, is the problem also happening with the [example of GLFW](https://github.com/go-gl/example) (this is to rule out whether it is imgui specific)?

**Comment by icexin** [comment]
Other examples of demo windows can normally close the program, only the console example will have this problem. I tried `example_sdl_opengl2`, `example_sdl_opengl3`, `example_glfw_opengl2`, all have this problem.

**Comment by icexin** [comment]
go version `go version go1.13 darwin/amd64`
OS version: `osx 10.13.6`
go env
```
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/fanbingxin/Library/Caches/go-build"
GOENV="/Users/fanbingxin/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/fanbingxin/develop/go"
GOPRIVATE=""
GOPROXY="https://goproxy.cn"
GOROOT="/Users/fanbingxin/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/Users/fanbingxin/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/fanbingxin/develop/go/src/github.com/inkyblackness/imgui-go-examples/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/10/l17n88zd1sj1csvh9_jx13_842c4yn/T/go-build154836210=/tmp/go-build -gno-record-gcc-switches -fno-common"
```

**Comment by dertseha** [comment]
I did another attempt at reproducing this - the animated video did indeed help to follow the steps. This crash was also happening on MS Windows.

It turns out that the test console is based on a static variable of function `ShowExampleAppConsole()`.
This is destructed upon exit of the application. And since this is in C-land, it happens (apparently) after the teardown of Go-land. During its destruction, it correctly wants to free any allocated memory, which is stored in a vector. This vector, in its index-operator, calls `IM_ASSERT` to verify the index is correct. And finally, this `IM_ASSERT` (in its previous override) always called into the Go-handler to check it. And this is bad when Go is already shut down.

**tl;dr**
The assertion callback is now two-staged: First the C-land checks whether the condition actually fails, and only calls into Go-land if it actually failed. Not only does this solve the current problem, it essentially also makes things faster, as all these assertions are now immediately handled in C-land.

Thank you for noticing this!
Code is changed in master already, new release is pending.
